Add MAC address support to find an interface name

sboeuf opened the pull request:
Instead of trying to guess which interface name will be created inside the VM, this patch allows any host user to give hyperstart the same MAC address that has been provided to the hypervisor.

Commit:
net: Parse MAC address from interface

In order to improve the way to find a network interface, we give the
possibility to provide a MAC address instead of the interface name.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

Commit:
net: Find network interface name from MAC address

In case of virtual machines, it is complicated to know what will
be the network interface name created by the hypervisor. But in
case we give a MAC address related to the TAP interface passed to
the hypervisor, we will be able to find this same address from
the agent running inside the VM.

That's why this patch provides a way to find a network interface
name from a MAC address, avoiding to try to predict what will be
the network interface name created.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
@sameo @dlespiau @jodh-intel @devimc @amshinde @mcastelino please take a look. It does not break CC 2.1 and it brings a new way (more reliable) to find the interface name. Eventually, it will allow some code removal inside cc-oci-runtime (2.1).

mcastelino commented on src/net.c:
```diff
@@ -777,6 +861,16 @@ static int hyper_setup_interface(struct rtnl_handle *rth,
+			" address %s\n", iface->mac_addr);
+		return -1;
+	}
+
```
@sboeuf can we add another check here to check is mac_address matches the one obtained using device name in case both device name and mac address are provided. It will offer a failsafe just in case something goes wrong on the host side.

sboeuf replied:
Yes I can do that in a follow up patch :)

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +760,82 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+			    strlen(ifa->ifa_name) + 1)) {
+			fprintf(stderr, "strncpy failed to copy interface"
+				" name: %s\n", strerror(errno));
+			goto err;
```
This will leak `sock`. In fact, it never seems to be closed.

sboeuf replied:
done

jodh-intel replied:
Have you re-pushed? I'm not seeing any changes?

sboeuf replied:
No no, the done flag is for myself so that I know I did the change :). I will push soon

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +760,82 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
 				IFLA_MTU);
 }
 
+static int hyper_get_iface_name_from_mac_addr(char *mac_addr, char **device)
```
- Missing checks on pointer parameters.
- `mac_address` can be `const` I think.
- I'd add a function header explaining what this function does, params, return value and also point out that it is limited to handling IPv4 MAC addresses.

sboeuf replied:
done

jodh-intel commented on src/net.c:
```diff
@@ -16,6 +21,8 @@
 
 #define UEVENT_BUFFER_SIZE 512
 
+#define MAC_ADDR_STR_SIZE 18
```
Could do with a comment showing format `XX:XX:XX:XX:XX:XX + 1` maybe. I'd consider naming it `IPV4_MAC_ADDR_STR_SIZE` to make it clearer that this is only for IPv4.

sboeuf replied:
done

Commit:
net: Check MAC address from network interface name

In case hyperstart receive both an interface name and a MAC
address, we want to make sure they are properly matching.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

Commit:
net: Factorize MAC address validation from a network interface name

Two functions were now using the same code to get a MAC address from
a network interface name, and validate it against an expected one.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
@jodh-intel @mcastelino I have updated it according to your comments.

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +759,202 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+err1:
+	close(sock);
+err:
+	return ret;
```
You don't really need `ret` here as it's always going to be `-1`. Alternatively, you could initialise `sock` and `tmp_mac_addr` which would allow you to simplify these errr handling at the end (by dropping 2 of the labels).

sboeuf replied:
Agreed.

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +759,202 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+err1:
+	freeifaddrs(ifaddr);
+err:
+	return ret;
```
Same comment as for `hyper_check_device_match_mac_addr()` - `ret` isn't really serving much of a purpose here.

sboeuf replied:
Agreed.

amshinde commented on src/net.c:
```diff
@@ -827,10 +837,12 @@ static int hyper_check_device_match_mac_addr(const char *device,
 
+	free(tmp_mac_addr);
 	close(sock);
 	return 0;
```
It would be better to assign ret = 0 and then fall through to return that instead of repeating the free and close lines

sboeuf replied:
Yes it would work the same, I just don't know what is preferred in this kind of case. @jodh-intel do you agree with @amshinde about this syntax ?

jodh-intel replied:
If we can avoid duplication - let's do that ;)

sboeuf commented:
@mcastelino @jodh-intel @amshinde I did the changes you suggested, let me know if you're okay with this patch now.
@sameo let me know your opinion on this one please.

devimc commented on src/net.c:
```diff
@@ -753,6 +759,195 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+
+		if (!strcasecmp(mac_addr, tmp_mac_addr)) {
+			free(tmp_mac_addr);
+			*device = strdup((const char*) ifr.ifr_name);
```
Hi @sboeuf 
I think before assign a new value to ```device``` you should validate that it is not null, if it is then free previous value in order to avoid memory leaks

sboeuf replied:
Yes sure I can do that.

sboeuf commented:
New update, I have addressed @devimc comment. I have tested it and it works perfectly with virtcontainers ;-)

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +759,198 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+	char *tmp_mac_addr = NULL;
+	int ret = -1;
+
+	if (!device && !strcmp(device, "")) {
```
This test and the `mac_addr` one below are incorrect and will likely cause a crash if `device == NULL`.

I'd tweak them to something like the following which saves the library call:

```
if (!device || !*device) {
```

sboeuf replied:
Yeah of course you're right, and I don't know why I used && instead of || ...

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +759,198 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+		goto err;
+	}
+
+	if (!strcmp(mac_addr, "")) {
```
This is fine, but why the different approach compared to `hyper_check_device_match_mac_addr()`?

sboeuf replied:
I have modified it in order to follow the same pattern than the previous function.

jodh-intel commented on src/net.c:
```diff
@@ -753,6 +759,198 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
+ * filled with the right network interface name. In case the function
+ * fails, it returns -1 and \p device will point to \c NULL.
+ */
+static int hyper_get_iface_name_from_mac_addr(const char *mac_addr,
```
Is there a reason the arguments to this function are the opposite way round compared to `hyper_check_device_match_mac_addr()`?

Also, both these functions seem to share common code so you might be able to refactor them into a single function and then `hyper_check_device_match_mac_addr()` and `hyper_get_iface_name_from_mac_addr()` can become "1-liners" that call this new function.

sboeuf replied:
Well I have reversed the arguments but there is a slight difference between those functions. First, we are not checking device the same way. And second, the strcasecmp() does not result in the same thing.
Moreover, I don't think it would be very efficient to have a function opening the socket and closing called at every iteration of the loop.
I think the best I can do is to factorize what is inside the loop. Do you agree ?

jodh-intel replied:
Yes - sounds good.

sboeuf commented:
@jodh-intel I updated the PR.

jodh-intel commented:
lgtm

sboeuf commented:
I don't know what is the trigger to decide if we merge it, I mean how many lgtm before we take it ?
@sameo, @mcastelino maybe take a look before we merge it.

mcastelino commented:
@sboeuf @sameo IMO we should merge this now. It will help in the case of DPDK and SRIOV also and reduce the interface specific setup complexity we have now.

dlespiau commented:
That's decided then! Makes a lot of sense to me to use the mac addresses and @egernst was complaining about the predictability of interface names this week.

@sboeuf could you make sure to create new issues for the work needed after this has landed:
  - update the hyperstart package and image version in cc-oci-runtime (could be assigned to @jcvenegas, @chavafg needs to test the new package with a full test run)
  - detail the work to be done in 2.1 to rely on that instead of interface names

sboeuf commented:
@dlespiau yes I'll do that

amshinde commented:
@dlespiau @sboeuf I can make the changes in 2.1 to support this once this lands in the image.

sboeuf commented:
@amshinde Great, that would allow to remove some code !

mcastelino commented:
@amshinde isn't this change backward compatible. So the current host side setup should work? And once it switches to macaddress based setup, things should become simpler and more uniform across both interface types as well as machine types.

sboeuf commented:
@mcastelino yes it is backward compatible

amshinde commented:
@mcastelino yes it is, it will help to simplify things as you said
